### PR TITLE
Add iree-reduce tool

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds an `iree-reduce` tool, a test case reduction tool which reduces the size of a given test case, producing the same error as the original test case.

The tool implementation is split into 2 parts: Framework and Strategies.
"Framework" is independent of IREE and implements the Delta Debugging Algorithm.
"Strategies" implements IREE-specific delta passes for test case reduction.

Currently, 3 delta passes are implemented:
1. ReduceLinalgOnTensorsDelta: Replaces linalg operations with a fill operation
2. ReduceFlowDispatchOperandToResultDelta: Replace the result of a flow.dispatch operation with it's operand, if they have the same shape
3. ReduceFlowDispatchResultBySplatDelta: Replace a flow.dispatch operation with a splat operation.

The delta passes are initial deltas that worked well. They can be upgraded as we need arises.

### CODE REVIEW

**Comment by Groverkss** [comment]
Currently, I'm looking for high level comments on the reducing infrastructure. Known things:

1. The build system is not very well written and piggybacks on the compiler driver. This should be fixed before landing.
2. The code needs to be more commented along with a "tutorial" on how to write a Delta strategy.

Other comments are very welcome on how we should build this infrastructure.

**Comment by Groverkss** [comment]
Reasons why I decided not to build on mlir-reduce:

1. mlir-reduce uses a reduction pattern interface. Reduction methods can be much more than just rewrite patterns and I find writing Delta passes much easier and more intuitive.
2. mlir-reduce is not the most flexible tool to do things. It also uses the Pass infrastructure, and I wanted more control than that.
3. Easier to iterate on IREE tool to show what works for us and later upstream it.

File: samples/iree-reduce/interesting.sh
Comment (benvanik) [inline samples/iree-reduce/interesting.sh]: you'll want to use vmvx as a target for things as CUDA is not always available - we try to keep most samples/tests (especially for tools like this that are generally useful) working in all configurations

File: compiler/src/iree/compiler/Reducer/CMakeLists.txt
Comment (ScottTodd) [inline compiler/src/iree/compiler/Reducer/CMakeLists.txt]: > Reasons why I decided not to build on mlir-reduce:
> 
> 1. mlir-reduce uses a reduction pattern interface. Reduction methods can be much more than just rewrite patterns and I find writing Delta passes much easier and more intuitive.
> 2. mlir-reduce is not the most flexible tool to do things. It also uses the Pass infrastructure, and I wanted more control than that.
> 3. Easier to iterate on IREE tool to show what works for us and later upstream it.

It looks like this tool has no dependencies on other IREE code as written. That's good for upstreaming prospects, but it does have me wondering if this code would have a better home in a fork of LLVM or some "sandbox" style repository (e.g. https://github.com/iree-org/iree-llvm-sandbox). Setting it up that way from the start would be a strong way to enforce that the tool does not grow any project-specific dependencies.

IDK what the sentiment is in upstream LLVM/MLIR, but maybe a `mlir-reduce-2` could even be created...?

**edit**: I see a bunch of IREE includes. Could probably add those to `DEPS`
```
#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
```
(would still be nice to have a clear boundary between what is generic and what is specific to this project)

File: compiler/src/iree/compiler/Reducer/Delta.cpp
Comment (ScottTodd) [inline compiler/src/iree/compiler/Reducer/Delta.cpp]: If you want to keep these logs, you could pass a `llvm::raw_ostream &os` in and send logs to that, rather than use `llvm::errs()` directly. Then, callers could choose what stream to use (if any).

File: compiler/src/iree/compiler/Reducer/CMakeLists.txt
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/CMakeLists.txt]: I split the Reducer into two, Framework and Strategies. The Framework part of it is isolated from IREE deps and can be readily upstreamed to MLIR. The Strategies on the other hand are IREE specific.

This should help in having the clear boundary.

File: compiler/src/iree/compiler/Reducer/Delta.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Delta.cpp]: I'll add a verbose option for these things I think.

File: compiler/src/iree/compiler/Reducer/CMakeLists.txt
Comment (stellaraccident) [reply compiler/src/iree/compiler/Reducer/CMakeLists.txt]: I think that IREE needs a tool like this desperately and whether or not there is a generalization or upstreaming strategy, that needs to follow, not lead.

I'm +1 on accepting it.

File: compiler/src/iree/compiler/Reducer/Framework/Delta.cpp
Comment (stellaraccident) [inline compiler/src/iree/compiler/Reducer/Framework/Delta.cpp]: Stray dumps here and throughout? I would have expected that if keeping these, you would just be emitting to debugOs (which may or may not be stderr).

File: compiler/src/iree/compiler/Reducer/Framework/WorkItem.h
Comment (stellaraccident) [inline compiler/src/iree/compiler/Reducer/Framework/WorkItem.h]: Note here and elsewhere: better to have a dedicated namespace (and can use compact form): `namespace mlir::iree_compiler::Reducer {`

File: compiler/src/iree/compiler/Reducer/Framework/WorkItem.h
Comment (stellaraccident) [inline compiler/src/iree/compiler/Reducer/Framework/WorkItem.h:24]: I would usually expect to see ownership conveyed here via `mlir::OwningOpReference<ModuleOp>` (unique_ptr semantics) or equiv. Since looks like would be a non-trivial refactor, I'm ok if you add a TODO to switch to that and get it in a followup.

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp
Comment (stellaraccident) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp]: sp: s/on/one/

File: samples/iree-reduce/interesting.sh
Comment (stellaraccident) [inline samples/iree-reduce/interesting.sh]: Maybe better to put this in a `test` directory.

Comment (stellaraccident) [reply samples/iree-reduce/interesting.sh]: Also, would be more portable if it was a python script (even if canonical usage is to use a shell script for brevity, python would be more testable)

File: samples/iree-reduce/output.mlir
Comment (stellaraccident) [inline samples/iree-reduce/output.mlir]: In a test directory

File: samples/iree-reduce/run_reducer.sh
Comment (stellaraccident) [inline samples/iree-reduce/run_reducer.sh]: Test dir

File: samples/iree-reduce/softmax.mlir
Comment (stellaraccident) [inline samples/iree-reduce/softmax.mlir]: test dir

File: compiler/src/iree/compiler/Reducer/Strategies/BUILD.bazel
Comment (benvanik) [inline compiler/src/iree/compiler/Reducer/Strategies/BUILD.bazel]: remove the comments as it prevents buildifier from sorting things (such segmentation usually bitrots fast as large scale changes/integrates are made and people don't know where to put things)

File: samples/iree-reduce/softmax.mlir
Comment (stellaraccident) [inline samples/iree-reduce/softmax.mlir]: We shouldn't have basic e2e tests that rely on a gpu. Can this be reformulated to work on CPU? Or maybe just added as a example.

File: compiler/src/iree/compiler/Reducer/Framework/README.md
Comment (benvanik) [inline compiler/src/iree/compiler/Reducer/Framework/README.md]: readmes as todos bitrot fast - keep these in issues or add more permanent documentation (what "framework" is, etc)

File: compiler/src/iree/compiler/Reducer/Framework/Oracle.h
Comment (benvanik) [inline compiler/src/iree/compiler/Reducer/Framework/Oracle.h]: use the llvm functionality for debugging instead of this - llvm::dbgs() and LLVM_DEBUG_TYPE/etc so that users can rely on standard tooling and compile out the debugging when not required

File: compiler/src/iree/compiler/Reducer/Framework/Delta.cpp
Comment (benvanik) [inline compiler/src/iree/compiler/Reducer/Framework/Delta.cpp]: dumping is very expensive and should be compiled out (LLVM_DEBUG/etc) - it also goes to llvm::dbgs() which may be different than debugOs (which is why it should either be llvm::dbgs() instead of passing in a different stream _or_ use print() that takes a stream)

**Review by stellaraccident — APPROVED** [review]
I'm ok on landing this as a basis of iterating. Left some nit-picky comments but the big ones are related to testing strategy: This needs tests that are running as part of the CI. I realize it can be a bit hard to get there in one step and I think the scripts you have are a starting point. Let's minimally move those to a test/ subdirectory and add a README describing how to run them with a note that getting them into the CI is being worked on. I can help you iterate to get them factored so we can enable them, and that will be easier to do without carrying the rest of the patch around.

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp
Comment (benvanik) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp]: naked TODOs should be rare and instead TODOs should be targeted at github users (yourself) or issues
https://google.github.io/styleguide/cppguide.html#TODO_Comments
(`// TODO(groverkss): do something.` or `// TODO(#issuenumber): do something.`)

File: samples/iree-reduce/run_reducer.sh
Comment (benvanik) [inline samples/iree-reduce/run_reducer.sh]: remove this script and add a README.md with this single command line instead

File: samples/iree-reduce/softmax.mlir
Comment (benvanik) [inline samples/iree-reduce/softmax.mlir]: use VMVX and drop the transform dialect stuff - CUDA is not always available (either built into the compiler or a CUDA device at runtime) and the transform dialect is not required to demonstrate the reducer in use

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by jpienaar** [comment]
Could you perhaps in the description say what kind of reductions it already supports? Perhaps a demo snippet showing what it could reduce and what not. Potentially more appropriate for a README here too.

(AFK but was going to check if this is on bytecode now, if the tool to execute could be fed in, e.g., that would have reduced your needing to rebase often section but could come at some other cost)

File: compiler/src/iree/compiler/API/Internal/IREEReduceToolEntryPoint.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/API/Internal/IREEReduceToolEntryPoint.cpp]: Is this to llvm::dbgs now?

File: compiler/src/iree/compiler/Reducer/Framework/ChunkManager.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Framework/ChunkManager.cpp]: Nit: prefer pre-increment unless post increment behavior is needed.

File: compiler/src/iree/compiler/Reducer/Framework/ChunkManager.h
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Framework/ChunkManager.h]: If you used a pair for these I think most of your operators is already provided. (Fine either way, but seeing as you already create pair for compare, seems improvement).

File: compiler/src/iree/compiler/Reducer/Framework/ChunkManager.h
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Framework/ChunkManager.h]: Prefer not to use unsigned to indicate non-negative

File: compiler/src/iree/compiler/Reducer/Framework/Delta.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Framework/Delta.cpp:19]: Failure is a bit tricky to interpret here. Especially as this is used to triage failing cases, so failure could represent success. Either define a new helper struct to make this clearer if the result matches some concept of interesting or not, or document here what failure means.

File: compiler/src/iree/compiler/Reducer/Framework/Delta.cpp
Comment (jpienaar) [reply compiler/src/iree/compiler/Reducer/Framework/Delta.cpp]: +1, capture in debug blocks and print to llvm::dbgs

File: compiler/src/iree/compiler/Reducer/Framework/WorkItem.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Framework/WorkItem.cpp]: Could this be added when needed?

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchOperandToResult.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchOperandToResult.cpp]: Could you add a function level comment?

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchOperandToResult.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchOperandToResult.cpp:50]: Note: this could quickly result in removing many interesting cases in general. I'd be cautious about simplifying without checking first (yes cost extra check but alternative is simplifying away the issue).

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp]: Ranked

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp:56]: So this is always all the results? (Trying to think if that is too conservative)

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp]: Could you expand on this? Why isn't it worth it?

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp
Comment (jpienaar) [inline compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp:29]: What is special about Linalg op here?

File: compiler/src/iree/compiler/Reducer/Framework/README.md
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Framework/README.md]: Currently removing the TODOs and will send them as issues instead.

File: compiler/src/iree/compiler/Reducer/Framework/WorkItem.h
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Framework/WorkItem.h:24]: Added a TODO for now

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchOperandToResult.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchOperandToResult.cpp:50]: I prefer the simplifications. I don't want to rely on dead-code to produce the interesting cases. I added "util.optimization_barrier" to other delta passes to preserve the dead code. Can be added to this also, but didn't really see the need right now.

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp:56]: Yes. I plan to add a delta pass which will start removing optimization barriers to check if we still can produce the test behavior we need.

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp:29]: Nothing really. I guess I could replace any tensor result with a fillop. Added a TODO for now, will add in subsequent patches.

File: compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp]: If there is no tensor input, then the operation is similar to a fill operation. This pass is supposed to reduce the number of linalg operations, not reduce the complexity of the linalg operation itself. Added inline documentation about this.

Comment (Groverkss) [reply compiler/src/iree/compiler/Reducer/Strategies/ReduceLinalgOnTensorsDelta.cpp]: On second thought, reducing might also help as it reduces it to an "identity" fill. Added a TODO about this.

File: samples/iree-reduce/run_reducer.sh
Comment (Groverkss) [reply samples/iree-reduce/run_reducer.sh:3]: Moved this as a filecheck RUN: command

**Comment by Groverkss** [comment]
I addressed all comments, as well as updated the patch description to better reflect today's state. Thank you for all the feedback on the patch!

The patch should be ready for another review pass on it.

File: samples/reducer/README.md
Comment (ScottTodd) [inline samples/reducer/README.md]: I don't see a `test/softmax.mlir` - should this be `test/reduce.mlir` instead? (Looks like it was renamed at some point)

File: samples/reducer/README.md
Comment (ScottTodd) [inline samples/reducer/README.md:13]: (No action needed, can follow-up when adding CI coverage)

Most other in-tree test scripts that do these sorts of "run command(s), check output or return value" operations are written in bash/shell, either in a standalone script file (that CI can run) or just inlined into a README. Here are a few examples:

* https://github.com/openxla/iree/blob/main/samples/variables_and_state/test.sh
* https://github.com/openxla/iree/blob/main/samples/vision_inference/README.md#instructions
* https://github.com/openxla/iree/blob/main/experimental/web/build_and_test_samples.sh

Bash is awkward compared to Python, but scripts written in it can be easier to splice together, copy step by step into a terminal session, and integrate with IDEs and CI systems.

That being said, I do recommend switching to Python when a certain complexity threshold is reached. Here's one example of that which also uses `subprocess.run`: https://github.com/openxla/iree/blob/main/samples/colab/test_notebooks.py. We also have more traditional Python tests in places like https://github.com/openxla/iree/tree/main/runtime/bindings/python/tests

File: samples/reducer/test/interesting.py
Comment (ScottTodd) [inline samples/reducer/test/interesting.py:9]: (No action needed, can follow-up when adding CI coverage)

Relating to my other comment about Bash vs Python for automation / test scripts, two patterns for running tools at some relative file location that we have are:

https://github.com/openxla/iree/blob/ef280a4d65e68b554ee81e8bbafa176188032c5c/samples/variables_and_state/test.sh#L15-L16 https://github.com/openxla/iree/blob/ef280a4d65e68b554ee81e8bbafa176188032c5c/samples/colab/test_notebooks.py#L36-L49

Those make the path traversal more explicitly dependent on the repo root or the script file location. We also have scripts that allow overriding the build directory location, rather than assuming it is always called just `build` (either as a child or sibling of the repo root).

File: samples/reducer/test/reduce.mlir
Comment (ScottTodd) [inline samples/reducer/test/reduce.mlir]: I'm trying to understand this file. Can you add some comments?

What is it testing? What should be getting reduced out of it? Is there an intentional bug in the file?

Ideally there would be enough context here for someone who does not work in this area to be able to update the file as changes are made to MLIR/IREE. If the file intentionally includes an error/bug, that could be particularly confusing to a maintainer to discover on their own without a comment.

File: tools/iree-reduce.cc
Comment (ScottTodd) [inline tools/iree-reduce.cc:1]: This file could use a comment explaining what the tool is designed to do, with a sample usage command.

Text could start as simple as what's in the PR description: "a test case reduction tool which reduces the size of a given test case, producing the same error as the original test case.".

I don't actually see comments like what I'm expecting in iree-opt, iree-run-module, iree-compile, etc. but here are some other examples:

* https://github.com/openxla/iree/blob/main/tools/iree-run-mlir-main.cc
* https://github.com/openxla/iree/blob/main/tools/iree-dump-module-main.c

File: compiler/src/iree/compiler/Reducer/BUILD.bazel
Comment (ScottTodd) [inline compiler/src/iree/compiler/Reducer/BUILD.bazel]: Can remove this `bool_flag` and `config_setting`. The file this is forked from used them here: https://github.com/openxla/iree/blob/ef280a4d65e68b554ee81e8bbafa176188032c5c/compiler/src/iree/compiler/Tools/BUILD.bazel#L212-L223

**Review by stellaraccident — APPROVED** [review]

**Comment by stellaraccident** [comment]
Looks like Ben's comments were addressed...

**Review by benvanik — APPROVED** [review]

**Comment by ScottTodd** [comment]
Nice! Thanks for addressing all of the comments :D

**Pull request merged**
